chain/ethereum/src/trigger.rs

```diff
@@ -229,11 +229,75 @@ impl ToAscPtr for MappingTrigger {
     }
 }
 
+#[derive(Clone, Debug)]
+pub enum LogRef {
+    FullLog(Arc<Log>, Option<Arc<TransactionReceipt>>),
+    LogPosition(usize, Arc<TransactionReceipt>),
+}
+
+impl LogRef {
+    pub fn log_index(&self) -> U256 {
+        match self {
+            LogRef::FullLog(log, _) => log.log_index.unwrap_or(U256::zero()),
+            LogRef::LogPosition(index, receipt) => receipt
+                .logs
+                .get(*index)
+                .unwrap()
+                .log_index
+                .unwrap_or(U256::zero()),
+        }
+    }
+
+    pub fn transaction_index(&self) -> Option<U64> {
+        match self {
+            LogRef::FullLog(log, _) => log.transaction_index,
+            LogRef::LogPosition(_, receipt) => Some(receipt.transaction_index),
+        }
+    }
+
+    fn transaction_hash(&self) -> Option<H256> {
+        match self {
+            LogRef::FullLog(log, _) => log.transaction_hash,
+            LogRef::LogPosition(_, receipt) => Some(receipt.transaction_hash),
+        }
+    }
```

Comment:
To be extra paranoid that the behaviour is equivalent between variants, I'd suggest we have `log_index`, `transaction_index`, `trasaction_hash` and all similar methods call `self.log()` first and then access the field.